File: 2019-10-04/src/gilded_rose.js
class Item {
    constructor(name, sellIn, quality){
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }
}

class Shop {
    constructor(items=[]){
        this.items = items;
    }
    decreaseQuality(i) {
        this.items[i].quality--;
    }
    increaseQuality(i) {
        if (this.items[i].quality < 50) {
            this.items[i].quality++;
        }
    }
    updateQuality() {
        for (var i = 0; i < this.items.length; i++) {
            if (this.items[i].name != 'Aged Brie' && this.items[i].name != 'Backstage passes to a TAFKAL80ETC concert') {
                if (this.items[i].quality > 0) {
                    if (this.items[i].name != 'Sulfuras, Hand of Ragnaros') {
                        this.decreaseQuality(i);
                    }
                }
            } else {
                this.increaseQuality(i);
                if (this.items[i].name == 'Backstage passes to a TAFKAL80ETC concert') {
                    if (this.items[i].sellIn < 11) {
                        this.increaseQuality(i);
                    }
                    if (this.items[i].sellIn < 6) {
                        this.increaseQuality(i);
                    }
                }
            }
            if (this.items[i].name != 'Sulfuras, Hand of Ragnaros') {
                this.items[i].sellIn = this.items[i].sellIn - 1;
            }
            if (this.items[i].sellIn < 0) {
                if (this.items[i].name != 'Aged Brie') {
                    if (this.items[i].name != 'Backstage passes to a TAFKAL80ETC concert') {
                        if (this.items[i].quality > 0) {
                            if (this.items[i].name != 'Sulfuras, Hand of Ragnaros') {
                                this.decreaseQuality(i);
                            }
                        }
                    } else {
                        this.items[i].quality = 0;
                    }
                } else {
                    this.increaseQuality(i);
                }
            }
        }

        return this.items;
    }
}
module.exports = {
    Item,
    Shop
}

File: 2023-04-28/ruby/gilded_rose.rb
class GildedRose

  def initialize(items)
    @items = items
  end

  def update_quality()
    @items.each do |i|
      item = Sellable.new(i)
      item.update_quality()
      if item.name != "Sulfuras, Hand of Ragnaros"
        item.sell_in = item.sell_in - 1
      end
      if item.name == "Backstage passes to a TAFKAL80ETC concert"
        if item.sell_in < 0
          item.quality = item.quality - item.quality
        end
      end
      if item.name == "Aged Brie"
        if item.sell_in < 0
          item.increase_quality()
        end
      end
      if item.name != "Aged Brie"
        if item.name != "Backstage passes to a TAFKAL80ETC concert"
          if item.sell_in < 0
            item.decrease_quality()
          end
        end
      end
    end
  end
end

class Item
  attr_accessor :name, :sell_in, :quality

  def initialize(name, sell_in, quality)
    @name = name
    @sell_in = sell_in
    @quality = quality
  end

  def to_s()
    "#{@name}, #{@sell_in}, #{@quality}"
  end
end

class Sellable
  attr_accessor :item

  def initialize(item)
    @item = item
  end

  def to_s()
    @item.to_s()
  end

  def name()
    @item.name()
  end

  def sell_in()
    @item.sell_in()
  end

  def quality()
    @item.quality()
  end

  def sell_in=(value)
    @item.sell_in = value
  end

  def quality=(value)
    @item.quality = value
  end

  def increase_quality()
    if self.quality < 50
      self.quality = self.quality + 1
    end
  end

  def decrease_quality()
    if self.quality > 0
      if self.name != "Sulfuras, Hand of Ragnaros"
        self.quality = self.quality - 1
      end
    end
  end

  def update_quality()
    if self.name == "Aged Brie"
      if self.quality < 50
        self.quality = self.quality + 1
      end
    end
    if self.name == "Backstage passes to a TAFKAL80ETC concert"
      if self.quality < 50
        self.quality = self.quality + 1
        if self.sell_in < 11
          self.increase_quality()
        end
        if self.sell_in < 6
          self.increase_quality()
        end
      end
    end
    if self.name != "Aged Brie" and self.name != "Backstage passes to a TAFKAL80ETC concert"
      self.decrease_quality()
    end
  end

end
